text/0000-generic-integers.md

```diff
@@ -0,0 +1,437 @@
+
+[`repr(u128)`]: https://github.com/rust-lang/rust/issues/56071
+[#3659]: https://github.com/rust-lang/rfcs/pull/3659
+
+## Documentation
+
+For now, additional `primitive.u` and `primitive.i` pages will be added to rustdoc, and the existing `uN` and `iN` pages will be left as-is. Eventually, if we're comfortable with it, we can remove the `uN` and `iN` pages entirely and use `primitive.i` and `primitive.u` as the single source of documentation for these types, plus the pages for `usize` and `isize`.
+
+There certainly is a precedent for this: as of right now, all of these pages share the same documentation, and the examples are modified to work for all of these types. Removing these separate pages would help remove documentation redundancy, although `usize` and `isize` would still have to be kept separate.
+
+Since the `uN` modules are currently deprecated, no `std::u` or `std::i` module should be added.
+
+## Possible lints
+
+Due to the presence of edge cases like `N = 0` and `N = 1`, it feels reasonable to add in a few lints to prevent people from doing silly things like:
+
+* casting anything to `u<0>` or `i<0>` (these are just the singleton zero, and so a cast is a meaningless operation)
+* coercing a literal integer to a generic integer (anything besides zero might overflow without a restriction on `N`, and once restrictions on `N` become possible, we can take that into account)
+
+Preferably, a lot of the lints surrounding generic integers should be added to clippy before being accepted into the compiler, since it's likely many of them will cause more headaches than they're worth. We can continue to investigate possible lints to avoid these errors.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This is a *big* change to the language, not to be taken lightly.
+
+One of the biggest drawbacks is that this *only* allows encoding the storage size of an integer, and doesn't let users refine integers further to only allow a range of values. For example, it would be excellent for an API taking a "percent" value to enforce that the value is between 0 and 100, but generic integers alone cannot do this.
+
+As an example, someone might end up using `u<7>` for a percent since it allows fewer extraneous values (`101..=127`) than `u<8>` (`101..=255`), although this actually just overcomplicates the code for little benefit, and may even make the performance worse.
+
+Overall, things have changed dramatically since [the last time this RFC was submitted][#2581]. Back then, const generics weren't even implemented in the compiler yet, but now, they're used throughout the Rust ecosystem. Additionally, it's clear that LLVM definitely supports generic integers to a reasonable extent, and languages like [Zig] and even [C][`_BitInt`] have implemented them. A lot of people think it's time to start considering them for real.
```

Comment:
I wouldn't say Zig has *generic* integers, it seems like they have arbitrarily-sized integers. Or is it possible to write code that is generic over the integer size?

Reply:
well actually you can

```zig
const std = @import("std");

fn U(comptime bits: u16) type {
    return @Type(std.builtin.Type {
        .Int = std.builtin.Type.Int {
            .signedness = std.builtin.Signedness.unsigned,
            .bits = bits,
        },
    });
}

pub fn main() !void {
    const a: U(2) = 1;
    const b: U(2) = 3;
    // const c: U(2) = 5; // error: type 'u2' cannot represent integer value '5'
    const d = std.math.maxInt(U(147));
    std.debug.print("a={}, b={}, d={}", .{ a, b, d });
    // a=1, b=3, d=178405961588244985132285746181186892047843327
}
```

Reply:
I guess that example is satisfactory enough, @RalfJung? Not really sure if it's worth the effort to clarify explicitly.

Reply:
Ah, neat.

C and LLVM only have concrete-width integers though, I think?

Reply:
I mean, C doesn't have generic *anything*, so, I guess you're right. Not 100% sure the distinction is worth it.

Reply:
Clang adds `_BitInt` to C++ as an extension and the number of bits can be generic: `template <size_t N> void example(_BitInt(N) a);` will deduce `N` but it only works on the actual `_BitInt` types, not just any signed integer type.